mathinspector/__init__.py

```diff
@@ -1,2 +1,217 @@
+from .console.builtin_print import builtin_print
+from .node import NodeEditor
+from .objects import ObjectTree
+from .modules import ModuleTree
+from .mainmenu import MainMenu
+from .project import SaveData
+from .animation import Animation
+from .widget import Notebook
+
+from .util import vdict, name_ext
+from .config import __version__, AUTOSAVE_PATH
+from .plot import plot
+from .doc import Help
+
+class mathinspector:
+	"""
+	Math Inspector: a visual programming environment for scientific computing
+	https://mathinspector.com
+
+	help
+	---
+	The mathinspector `help` function has been designed to override the builtin python `help` function.  This makes it easy to write code that displays documentation for your own projects with the mathinspector doc browser.
+
+	These can be viewed using the help function.  For example, to view this documentation in the mathinspector doc browser, use the command
+
+	>>> import mathinspector
+	>>> from mathinspector import help
+	>>> help(mathinspector)
+
+	For more information about any of the items which follow, see the corresponding docstrings for that item.  If you are using the doc browser to view this documentation, the sidebar contains a list of everything this module contains, and you can click on those items to learn more.  All code examples can be clicked on, and they will be executed and the output printed to the command line.
+
+	main
+	---
+	This will launch the mathinspector app
+
+	>>> import mathinspector
+	>>> mathinspector.main()
+
+	See the documentation for the main function to understand the structure of the arguments it accepts
+
+	plot
+	---
+	The mathinspector plotting library was designed to update and replace the functionality in matplotlib.
+
+	This will launch the mathinspector plotting window.
+
+	>>> import mathinspector
+	>>> mathinspector.plot(1,2,3)
+
+	The way you format data you pass as arguments to the plot function will determine what is plotted.  It will automatically detect when the input is 2-dimensional or 3-dimensional, and treats tuple's as points and lists & arrays as lines.  See the documentation for the plot function for a complete list of accepted argument types.
+
+	vdict
+	---
+	This module also has the class vdict, which stands for virtual dictionary.  It's a class which extends dict
+	and has callbacks for getting, setting, and deleting items.
+
+	>>> import mathinspector
+	>>> A = mathinspector.vdict(setitem=lambda key, value: print(key))
+	>>> A["x"] = 1
+	    x
+
+	"""
+	def __dir__(self):
+		return ALLOWED_NAMES
+
+	def __getattr__(self, name):
+		if name in ALLOWED_NAMES:
+			return globals()[name]
+		raise AttributeError(f"module {__name__} has no attribute {name}")
+
+ALLOWED_NAMES = [
+	"__builtins__",
+	"__cached__",
+	"__doc__",
+	"__file__",
+	"__loader__",
+	"__name__",
+	"__package__",
+	"__path__",
+	"__spec__",
+	"__version__",
+	"animate",
+	"main",
+	"plot",
+	"vdict",
+	"help",
+]
+
+help = Help()
+sys.modules[__name__] = mathinspector()
```

Comment:
zoïks

Reply:
> super hacky way to prevent importing other things... The normal way to do this is to move __init__ to app and only import things in __init__ which you want exposed.

I tried this, and when I imported `main` it imported the entire util module and, whenever I imported anything it flooded the result of `dir(mathinspector)` with tons of stuff.  My understanding after researching this was there are 2 schools of thought

- just let everything get imported
- use the super hacky way to do it that used to be "official" way to do it (according to what i read on SO)

Was I doing it wrong, and `dir(mathinspector)` should be clean after `from .app import main`?  Or are you of the first school of thought?  I get that its super hacky, but I felt it was (partially) justifiable.  The negative consequences are unclear to me at this time, curious what could go wrong?

> someone calling mathinspector --help on the CLI does not need the app to run at all
You probably saw this, but it doesn't run the app in that scenario (the 2nd argument in the constructor of the `Help` class is the app, and when this is blank, it only opens the doc window.

> Is there another reason why you do this?
Good question.  I haven't completely made up my mind about `app.py`.  My thinking was:

- I want to minimize the number of files in the root of the mathinsector/ folder
    * I think this helps people who want to read the code and get involved figure things out a lot faster when done right
- the `__init__` files big job is to be responsible for calling `main`, so that function should be defined in that file
- The App class is only used a single time in the entire code base, in main
- Blocks of code which are only used for one thing with other parts of code should be next to each other in the same file unless that file gets too big
- For a short class like App, it's not helping code readability to move it into its own file
    * its actually hurting it, since it's hiding something that you need to know about to understand how __init__ works

You may have noticed I care a lot about code purity, naming conventions, and file/dir structure.  In this case, I am not completely convinced keeping App in it's own file `app` is better or worse than `__init__`  For people just starting to read the code base, it saves them a click and a moment of wondering where App came from.  Small wins like that can add up to big improvements in overall code readability and quality.

> I'm going to close my eyes over the other dark magicke of redefining help(), print(), sanity(), etc..

I have been holding my nose on that lol.  Here is how I feel about global variables.  If you know what you are doing, then they are a tool like everything else.  I am not afraid of using global variables, I use them when they are the right tool for the job.  In this case, it looks like this is a good use a global variables.

Reply:
tl;dr Is there a way to get the identical output for `dir(mathinspector)` without doing the super hacky thing?

Reply:
Ok I removed the sys.modules stuff https://github.com/MathInspector/MathInspector/pull/38/commits/8c482ca286daee9e7398ce2b4b0d28b1ab931232

It's probably better to have it count as a module when you do things like `inspect.ismodule(mathinspector)` for one thing.  It's no big deal if someone wants to use undocumented parts of the mathinspector api i guess, no reason to lock it down with the sys.modules hack